Add additional text info to content picker field, so user can distigu…

MikeKry opened the pull request:
Added culture and autoroute (if filled) to content item picker, so user can easily pick content items. Using only display name has become a problem for multilingual sites with page named e.g. "blog" in multiple languages.

![image](https://user-images.githubusercontent.com/52829889/104335508-b7100a80-54f3-11eb-98a0-e6d589359412.png)

MikeKry commented:
@deanmarcussen 

I am not sure, if I should assign somebody to look at this, or what is the exact process with PRs? :)

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.ContentFields/ContentItemDisplayExtensions.cs:
```diff
@@ -0,0 +1,39 @@
+        public static async Task<string> GetContentItemRouteAndCulture(this ContentItem contentItem, IContentManager contentManager, IStringLocalizer S)
+        {
+            var displayText = string.Empty;
+            var additionalParts = new List<string>();
```
this would be a `StringBuilder` from the string builder pool

MikeKry replied:
moved and reflected in [ContentPickerContentHandler.cs]

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.ContentFields/ContentItemDisplayExtensions.cs:
```diff
@@ -0,0 +1,39 @@
+            }
+
+            // add autoroute path to display text
+            var routeAspect = await contentManager.PopulateAspectAsync<RouteHandlerAspect>(contentItem);
```
don't use this aspect get the path. It's not pathbase / tenant aware. this is used for contained routing.
You need the url helper, and `ContentItemMetadata.DisplayRouteValues` to generate a valid path.

deanmarcussen reviewed (commented):
Did you open an issue for this @MikeKry ?

Normally we start with issues, offer design guidance, (triage is on Thursday at 10 (PST?)), and then look at PRs.

However, for this one. Lets look at it the issue.

You want more metadata in the content picker?

I like the principle of the idea.

This sounds like an aspect.
We used to have a `DisplayText` property on the `ContentItemMetadata` which allowed you to manipulate the display text's for things like this. Unfortunately it's been obsoleted, and replaced in favour of just `ContentItem.DisplayText`

Maybe we could have a `AdditionalDisplayText` property there.

Here the problem is you've tied it to Culture, and Route. 
What if I want Alias? And what if I don't want Route? 
Or don't want any of them?

The routes on some of my sites are far too long to be displayed in the content picker list, for example.

MikeKry commented:
@sebastienros 

Ok, thanks for reply. So everything should always start with issue/feature request?  I have created it backwards for now #8374

From your comment I see the issue, that not everyone would like this format of display, I think it can use similiar setup as Autoroute part => liquid syntax so you can define anything you want here, but I am not sure, if it is possible within contentpicker context.

MikeKry commented:
Updates based on 8374

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentPickerResultProvider.cs:
```diff
@@ -7,6 +7,8 @@ namespace OrchardCore.ContentManagement
     {
         string Name { get; }
+
+        Task<string> GetContentPickerItemDescription(ContentItem contentItem, string pattern, string defaultValue);
```
Don't change this interface, or do anything to do with descriptions here.

MikeKry replied:
removed title and description from this interface

deanmarcussen reviewed (commented):
Concept looks good

Suggest:
`ContentPickerAspect` and using the aspect system from the content manager to generate the code you have put in the result provider.

To implement, make a `ContentPickerAspect` class w/ `DisplayText` and `Description`

implement a `ContentHandlerBase` and override `GetContentItemAspectAsync` `.ForAspect<ContentPickerAspect` and do the rendering work there.

Then you can safely inject it in either the driver, or the result provider as required.

MikeKry commented:
@deanmarcussen 

I have finally looked at aspect system and reworked this to aspects. Hopefully it is ok now - tested and still works, removed bunch of unnecessary code in existing implementations.

2 problems I met and not sure they are solved right:
a) in ContentPickerAspect there are content part settings passed as object, because there can't be reference to ContentPickerFieldSettings - did not want to move settings or aspect, as aspect should be referenced from anywhere.
b) in ContentPickerAspect - passing CultureInfo - > I cannot use ContentManager in ContentHandlerBase as it is circular reference, so it is passed into aspect, is it ok?

MikeKry commented:
@deanmarcussen 
does it give sense to you like that? there are some merge conflicts now but I can update this branch with master if its ok with you

deanmarcussen commented:
Concept looks good @MikeKry. Would suggest reading the previous pr review

MikeKry commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/ContentPickerFieldDisplayDriver.cs:
```diff
@@ -19,13 +19,16 @@ public class ContentPickerFieldDisplayDriver : ContentFieldDisplayDriver<Content
             IContentManager contentManager,
-            IStringLocalizer<ContentPickerFieldDisplayDriver> localizer)
+            IStringLocalizer<ContentPickerFieldDisplayDriver> localizer,
+            IContentPickerResultProvider contentPickerResultProvider)
```
removed

MikeKry commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/ContentPickerFieldDisplayDriver.cs:
```diff
@@ -42,7 +45,7 @@ public override IDisplayResult Display(ContentPickerField field, BuildFieldDispl
         public override IDisplayResult Edit(ContentPickerField field, BuildFieldEditorContext context)
         {
-            return Initialize<EditContentPickerFieldViewModel>(GetEditorShapeType(context), async model =>
+            return Initialize(GetEditorShapeType(context), (Func<EditContentPickerFieldViewModel, ValueTask>)(async model =>
```
reverted

MikeKry commented:
@deanmarcussen 

I looked at your comments when I reworked this and I think I have reflected all your comments in https://github.com/OrchardCMS/OrchardCore/pull/8280/commits/72661068bc33f4788ed78d478b54f91bc543beb6

I have gone through them again and I think everything should be ok, so I commented and resolved them.

MikeKry commented:
@Skrypt 
Ok, I was out for a while, but I have merged it now and tested - still working ok.

MikeKry commented:
@Skrypt  
Just a question - does "Needs triage" mean anything for me, or is it just waiting for some group review? I am not very familliar with how this triage etc. works.

Skrypt commented:
Needs triage means that we need to assign it to a milestone from a Thursday triage meeting. I use also this to make it prioritary over other PRs so that we can review it.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneContentPickerResultProvider.cs:
```diff
@@ -61,7 +61,7 @@ public async Task<IEnumerable<ContentPickerResult>> Search(ContentPickerSearchCo
                         ContentItemId = doc.GetField("ContentItemId").GetStringValue(),
                         DisplayText = doc.GetField("Content.ContentItem.DisplayText").GetStringValue(),
-                        HasPublished = doc.GetField("Content.ContentItem.Published").GetStringValue() == "true" ? true : false
+                        HasPublished = doc.GetField("Content.ContentItem.Published").GetStringValue() == "true" ? true : false,
```
```suggestion
                        HasPublished = doc.GetField("Content.ContentItem.Published").GetStringValue() == "true" ? true : false
```

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/ContentPickerFieldSettings.Edit.cshtml:
```diff
@@ -53,6 +61,22 @@
+        <label asp-for="TitlePattern">@T["Title Pattern"]</label>
+        <textarea asp-for="TitlePattern" rows="5" class="form-control"></textarea>
+    </div>
+    <span class="hint">@T["The pattern used to render the custom content item title in content picker. DisplayText is used when empty is empty. With Liquid support."]</span>
```
```suggestion
    <span class="hint">@T["The pattern used to render the custom content item title in content picker. DisplayText is used when title is empty. With Liquid support."]</span>
```

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Handlers/ContentPickerContentHandler.cs:
```diff
@@ -0,0 +1,53 @@
+            });
+        }
+
+        private async Task<string> GetContentPickerItemExtendedDescription(ContentItem contentItem, CultureInfo culture, string pattern, string defaultValue)
```
```suggestion
        private async Task<string> GetContentPickerItemExtendedValue(ContentItem contentItem, CultureInfo culture, string pattern, string defaultValue)
```

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Handlers/ContentPickerContentHandler.cs:
```diff
@@ -0,0 +1,53 @@
+
+        private async Task<string> GetContentPickerItemExtendedDescription(ContentItem contentItem, CultureInfo culture, string pattern, string defaultValue)
+        {
+            var description = defaultValue;
```
```suggestion
            var result = defaultValue;
```

Skrypt replied:
Here, I don't think you need to do this. This adds string allocation for nothing. You could also use defaultValue and morph its value and return it.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Handlers/ContentPickerContentHandler.cs:
```diff
@@ -0,0 +1,53 @@
+                
+                using (CultureScope.Create(culture))
+                {
+                    description = await _templateManager.RenderStringAsync(pattern, NullEncoder.Default, contentItem,
```
```suggestion
                    result = await _templateManager.RenderStringAsync(pattern, NullEncoder.Default, contentItem,
```

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Handlers/ContentPickerContentHandler.cs:
```diff
@@ -0,0 +1,53 @@
+                }
+            }
+
+            return description;
```
```suggestion
            return result;
```

Skrypt commented on src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentPickerResultProvider.cs:
```diff
@@ -21,6 +21,7 @@ public class ContentPickerSearchContext
     public class ContentPickerResult
```
Can you move and extract these classes into separate files? I'm pretty sure they don't belong in the abstraction project.

MikeKry commented:
@Skrypt  

Implemented suggested changes, 

Merged main into this branch but **I still need to check tomorrow if it did not break anything** - kinda lot changes

MikeKry commented:
@Skrypt 

Tested, fixed codemirror. It is working fine again with the new changes. Maybe it should be reviewed by @MikeAlhayek, because there were many a lot of conflicts with https://github.com/OrchardCMS/OrchardCore/pull/12103

MikeKry commented:
Is it possible, that this will make it to version 1.5?

sebastienros commented:
Sorry for the late reply, and for what's coming ;)

I think it can be much simpler with a single setting that contains the liquid template to use for the description of the content item in this list. Similarly to autoroute paths, with a default value that is just `{{ ContentItem | display_text }}`. Then users can customize each field to display the culture, type, ... without requiring any custom code. You did that part in this PR. I don't think we need to be able to do it by code.

NB: the way you were passing settings and culture to the aspect population made me realize this. Actually each implementation of the handler to populate the aspect should have resolved the culture if they needed it. Same for the settings.

MikeKry commented:
@sebastienros 

Not sure what you mean, because Autoroute paths generation also pass culture aspect to liquid templates (and imho it can't be generated without manually getting culture aspect). 

It can be seen here: https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutoroutePartHandler.cs#L406

Also there is no custom code needed from developer, title and description can resolve any liquid expression same as autoroute part.

But it is possible I misunderstood what you mean..? Maybe I would need a little more explanation.

sebastienros commented:
Correct, you will need to get the culture from the aspect, but then the template is sufficient, something like this:

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutoroutePartHandler.cs#L400-L409

MikeKry commented:
@sebastienros 

So you mean, that new ContentPickerAspect is not needed? That is ok and I can change it, but feels like step-back to state before this commit https://github.com/OrchardCMS/OrchardCore/pull/8280/commits/72661068bc33f4788ed78d478b54f91bc543beb6#diff-c12e6f2791029f70fa686c607691139163bfd007933317dfb86d22c3122e671f

Where @deanmarcussen suggested, that it should rather be an aspect. Are you sure that we want to change it back? I have no problem in doing it, but I admit that I do not understand this concept of aspects and when to use them in OC :).

If so, please suggest me where to place function for resolving liquid expression for description/title. I am not sure whether it should be placed in some handler or service.

MikeKry commented:
@sebastienros 

What is your decision, should I revert it or no? I would like to finish this as it is open for 1 year already :)

MikeAlhayek commented:
@MikeKry can you please fix the conflict?

MikeKry commented:
@MikeAlhayek 

Sure, I pushed merge into my branch. Is there possibility, that it will be merged any time soon? It starts to be very uncomfortable using content picker with growing number of content items.

MikeKry commented:
I see that release 1.6 is already merged, but this PR did not make it through.. will it ever be merged?

sebastienros commented:
Tasks to unblock the PR
- remove the aspect
- add a setting that defines the text to display in the list
  - either use an extra checkbox to define an optional setting that will then be used
  - use the setting all the time, and the default value is the liquid text to get the display name of the content item (backward compatibility)

sebastienros commented:
I would go with a single field right now, not changing how it looks in the UI. We can always add an extra one later on if someone finds a reason to. The view model should not change then, and the view either.

MikeAlhayek commented:
> Tasks to unblock the PR
> 
>     * remove the aspect
> 
>     * add a setting that defines the text to display in the list
>       
>       * either use an extra checkbox to define an optional setting that will then be used
>       * use the setting all the time, and the default value is the liquid text to get the display name of the content item (backward compatibility)

@MikeKry are you able to do this so it can be unblocked and hopefully merged before 1.7 is released?

MikeKry commented:
@sebastienros  @MikeAlhayek 

Ok, I have removed aspect and field for description. I have added default value for backwards compatibility. 

![image](https://github.com/OrchardCMS/OrchardCore/assets/52829889/5e16e7b7-3841-458c-892d-6c0b5846286b)

![image](https://github.com/OrchardCMS/OrchardCore/assets/52829889/da3806c9-e314-49c0-93ff-7f871907168c)

![image](https://github.com/OrchardCMS/OrchardCore/assets/52829889/fc72e1fd-faf6-4be3-9640-a88eb6522bda)

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/ContentPickerFieldSettings.Edit.cshtml:
```diff
@@ -1,6 +1,16 @@
+<script asp-name="codemirror-addon-mode-simple" at="Foot"></script>
+<script asp-name="codemirror-addon-mode-multiplex" at="Foot"></script>
+<script asp-name="codemirror-mode-xml" at="Foot"></script>
+
```
please remove extra space.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettingsViewModel.cs:
```diff
@@ -15,4 +15,5 @@ public class ContentPickerFieldSettingsViewModel
     public bool Multiple { get; set; }
 
     public string[] DisplayedContentTypes { get; set; } = Array.Empty<string>();
+    public string TitlePattern { get; set; } = "{{ Model.ContentItem.DisplayText }}";
```
please add a new line between line 17-18

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettingsDriver.cs:
```diff
@@ -59,12 +65,30 @@ public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinitio
+            if (!string.IsNullOrEmpty(model.TitlePattern) && !_templateManager.Validate(model.TitlePattern, out var titleErrors))
+            {
+                context.Updater.ModelState.AddModelError(nameof(model.TitlePattern), S["TitlePattern doesn't contain a valid Liquid expression. Details: {0}", string.Join(" ", titleErrors)]);
+                isValid = false;
```
Please re-label `S["TitlePattern doesn't contain a valid Liquid expression. Details: {0}"` to `S["Title Pattern does not contain a valid Liquid expression. Details: {0}"`

The `return false;`

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettingsDriver.cs:
```diff
@@ -59,12 +65,30 @@ public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinitio
+                isValid = false;
+            }
+
+            return isValid;
```
return true instead of using isValid;

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Services/DefaultContentPickerResultProvider.cs:
```diff
@@ -48,15 +59,21 @@ public async Task<IEnumerable<ContentPickerResult>> Search(ContentPickerSearchCo
             {
-                results.Add(new ContentPickerResult
+                var cultureAspect = await _contentManager.PopulateAspectAsync(contentItem, new CultureAspect());
+                using (CultureScope.Create(cultureAspect.Culture))
```
use a semicolon `;` at the end of the using line and drop the `{  }` to save indentation. 
```
using CultureScope.Create(cultureAspect.Culture);
```

MikeAlhayek replied:
@MikeKry this one please

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettings.cs:
```diff
@@ -10,5 +11,10 @@ public class ContentPickerFieldSettings
+        /// <summary>
+        /// The pattern used to build custom title
+        /// </summary>
+        public string TitlePattern { get; set; } = "{{ Model.ContentItem.DisplayText }}";
```
add attribute [DefaultValue("{{ Model.ContentItem.DisplayText }}")] to the new property since it has a custom default value.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettingsDriver.cs:
```diff
@@ -59,12 +65,30 @@ public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinitio
-                context.Builder.WithSettings(settings);
+                var isValid = ValidateTextPatterns(context, model);
+
+                if (isValid)
```
maybe `if(IsValidateTextPatterns(context, model))` instead?

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/ContentPickerFieldSettingsDriver.cs:
```diff
@@ -59,12 +65,30 @@ public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinitio
             return Edit(partFieldDefinition);
         }
 
+        private bool ValidateTextPatterns(UpdatePartFieldEditorContext context, ContentPickerFieldSettingsViewModel model)
```
maybe rename the method do `IsValidTitlePattern`

MikeAlhayek reviewed (commented):
@MikeKry can you confirm that HTML is supported in the picker? For example, if I provide the following pattern, will the HTML show up at text or will we see my custom HTML?

```
<a href="/Admin/Contents/ContentItems/{{ Model.ContentItem.ContentItemId }}/Display" target="_blank">{{ Model.ContentItem.DisplayText }}</a>
```

MikeKry commented:
@MikeAlhayek 

no, html is not rendered. I don't think, it would be good to allow html inside selector, it could break html structure etc.
![image](https://github.com/OrchardCMS/OrchardCore/assets/52829889/1e59a440-6144-4481-a660-5f355be47622)


All but one suggestions were applied to code in new commit.

MikeAlhayek commented:
@MikeAlhayek I think allowing HTML will solve for other scenarios. and if you are providing a custom template contained HTML, you should know what you are doing and the app should allow it.

If you don't mind, please add HTML support in the selector.

MikeKry commented:
@MikeAlhayek 

Okay, would you like html to be rendered both while selecting and also when items are already selected? 

In example with hyperlinks it can cause redirection when clicking on item to select (but that may be also somehow controlled by html that is inserted)

![image](https://github.com/OrchardCMS/OrchardCore/assets/52829889/7284d52a-150f-4816-b4d9-3705792c24c2)

MikeKry commented:
I have pushed version with both variants showing rendered html

MikeAlhayek commented:
Humm, I think in the suggestion we only show the Display text, but the selected options should support HTML. Like if the user want to add a link to edit the selected content item or view the selected content item.

MikeKry commented:
well I think that would be very complex solution.. Stripping html when searching is possible but would lead to nonsense results and using only display text would dump original intention of this pull request.

Main intention why I suggested this change to content picker is to enable quicker and easier searching - using display text only when searching would change nothing in this usecase.

MikeAlhayek commented:
I understand. I think it would be great if we have a solution for this. See https://github.com/OrchardCMS/OrchardCore/pull/13764 adding HTML will solve that issue as well.

MikeKry commented:
@MikeAlhayek  

I have reverted html rendering from this feature, as I think that solution suggested in #13764 is way better than allowing custom html.

If we would want to go through with allowing custom html, then I think it would be necessary to create another liquid field with definition for display of selected items. Then it would give sense to use it for buttons etc. but I think it is a bit overkill.

MikeAlhayek commented:
@sebastienros I think this one is ready, can we triage it?
